omod/src/main/java/org/openmrs/module/pihcore/fragment/controller/dashboardwidgets/HivAlertsFragmentController.java

```diff
@@ -0,0 +1,147 @@
+package org.openmrs.module.pihcore.fragment.controller.dashboardwidgets;
+
+import org.joda.time.DateTime;
+import org.openmrs.Concept;
+import org.openmrs.Encounter;
+import org.openmrs.EncounterType;
+import org.openmrs.Obs;
+import org.openmrs.Patient;
+import org.openmrs.api.ConceptService;
+import org.openmrs.api.EncounterService;
+import org.openmrs.api.ObsService;
+import org.openmrs.api.PatientService;
+import org.openmrs.api.context.Context;
+import org.openmrs.module.emrapi.patient.PatientDomainWrapper;
+import org.openmrs.module.pihcore.PihEmrConfigConstants;
+import org.openmrs.parameter.EncounterSearchCriteriaBuilder;
+import org.openmrs.ui.framework.UiUtils;
+import org.openmrs.ui.framework.annotation.InjectBeans;
+import org.openmrs.ui.framework.annotation.SpringBean;
+import org.openmrs.ui.framework.fragment.FragmentConfiguration;
+import org.openmrs.ui.framework.fragment.FragmentModel;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+
+public class HivAlertsFragmentController {
+
+	private final Logger log = LoggerFactory.getLogger(getClass());
+
+	public void controller(@SpringBean("obsService") ObsService obsService,
+							@SpringBean("conceptService") ConceptService conceptService,
+							@SpringBean("encounterService") EncounterService encounterService,
+							@SpringBean("patientService") PatientService patientService,
+						   @InjectBeans PatientDomainWrapper patientWrapper,
+							UiUtils ui,
+							FragmentConfiguration config,
+							FragmentModel model) {
+
+		// pretty much stole all this from the coreapps dashboard widgets controller... could this be simpler or extracted out somewhere?
+		Object patient = null;
+		patient = config.get("patient");
+		if (patient == null ) {
+			patient = config.get("patientId");
+		}
+
+		if (patient != null) {
+			if (patient instanceof Patient) {
+				patientWrapper.setPatient((Patient) patient);
+			} else if (patient instanceof PatientDomainWrapper) {
+				patientWrapper = (PatientDomainWrapper) patient;
+			} else if (patient instanceof Integer) {
+				// assume we have patientId
+				patientWrapper.setPatient(Context.getPatientService().getPatient((Integer) patient));
+			} else {
+				throw new IllegalArgumentException("Patient must be of type Patient or PatientDomainWrapper");
+			}
+		}
+
+		generateLateForMedPickupAlert(model, patientWrapper, encounterService, conceptService, obsService);
+		generateLateForAppointmentAlert(model, patientWrapper, encounterService, conceptService,obsService);
+
+	}
+
+	private void generateLateForMedPickupAlert(FragmentModel model, PatientDomainWrapper patientWrapper, EncounterService encounterService, ConceptService conceptService, ObsService obsService) {
```

Comment:
I won't pick through all of this logic yet.  My main thought is that this logic should not be hardcoded in here like this in a way that will require programming to change and which is opaque to the clinical and implementation team as to how it is calculated.  

A challenge with the HIV EMR is often that the definition for things like "lost to followup", "late for visit", "late for med pickup", etc. change very frequently as protocols or CDC/Pepfar definitions change.  At the same time, for every alert on the dashboard, there is likely a corresponding set of other tools that need the same criteria.  These include:

* Give me a data export of all patients who are currently late for their med pickup
* Build me an indicator around the number of patients late for their med pickup
* Give me an interactive patient list that shows me all patients who are late for their appointment

I wonder if an alternative here would be to have a set of report or data set definitions, defined via configuration, that this alert controller could iterate over and use to determine which alerts to display for a given patient.

Thoughts @mogoodrich ?